Client/Frontend/Widgets/AutocompleteTextField.swift

```diff
@@ -86,27 +91,50 @@ class AutocompleteTextField: UITextField, UITextFieldDelegate {
+    // get the new autocompletion.
     func textField(textField: UITextField, shouldChangeCharactersInRange range: NSRange, replacementString string: String) -> Bool {
-        if completionActive && string.isEmpty {
-            // Characters are being deleted, so clear the autocompletion, but don't change the text.
-            removeCompletion()
-            return false
+        if completionActive {
+            if string.isEmpty {
+                // Characters are being deleted, so clear the autocompletion, but don't change the text.
+                removeCompletion()
+                return false
+            }
+            removeCompletionIfRequiredForEnteredString(string)
         }
-
         return true
     }
 
+    private func removeCompletionIfRequiredForEnteredString(string: String) {
+        // If user-entered text does not start with previous suggestion then remove the completion.
+        let actualEnteredString = text.substringToIndex(advance(text.startIndex, enteredTextLength, text.endIndex)) + string
+        if !previousSuggestion.startsWith(actualEnteredString) {
+            removeCompletion()
+        }
+        enteredTextLength += count(string)
+    }
+
     func setAutocompleteSuggestion(suggestion: String?) {
-        if let suggestion = suggestion where editing && canAutocomplete {
+        // Setting the autocomplete suggestion during multi-stage input will break the session since the text
+        // is not fully entered. If `markedTextRange` is nil, that means the multi-stage input is complete, so
+        // it's safe to append the suggestion.
+        if let suggestion = suggestion where editing && canAutocomplete && markedTextRange == nil {
             // Check that the length of the entered text is shorter than the length of the suggestion.
             // This ensures that completionActive is true only if there are remaining characters to
             // suggest (which will suppress the caret).
-            if suggestion.startsWith(text) && count(text) < count(suggestion) {
-                let endingString = suggestion.substringFromIndex(advance(suggestion.startIndex, count(self.text)))
+            let enteredText = text.substringToIndex(advance(text.startIndex, enteredTextLength, text.endIndex))
+            if suggestion.startsWith(enteredText) && count(enteredText) < count(suggestion) {
+                let endingString = suggestion.substringFromIndex(advance(suggestion.startIndex, count(enteredText)))
                 let completedAndMarkedString = NSMutableAttributedString(string: suggestion)
                 completedAndMarkedString.addAttribute(NSBackgroundColorAttributeName, value: AutocompleteTextFieldUX.HighlightColor, range: NSMakeRange(enteredTextLength, count(endingString)))
                 attributedText = completedAndMarkedString
                 completionActive = true
+                previousSuggestion = suggestion
```

Comment:
How about we return early here and remove the `if else` below? Then, at the very bottom of this function outside of all the `if` statements, you do `removeCompletion`? That way, if the setting the suggestion fails for any reason, we always clear the existing completion if there is one.